packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/PathUtils.java

```diff
@@ -0,0 +1,278 @@
+  static final String providersDownloadsDocuments = "com.android.providers.downloads.documents";
+  static final String providersMediaDocuments = "com.android.providers.media.documents";
+  /**
+   * Given a projection, returns the file name for the uri using the content resolver to avoid
+   * unwanted parsing.
+   */
+  public static String getFileName(Uri uri, @NonNull Context context, String[] projection) {
+    Cursor returnCursor = context.getContentResolver().query(uri, projection, null, null, null);
+
+    int nameIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
+    returnCursor.moveToFirst();
+    String name = returnCursor.getString(nameIndex);
+    returnCursor.close();
+    return name;
+  }
+
+  @VisibleForTesting
+  ArrayList<String> copyFilesToInternalStorage(ArrayList<Uri> uri, Context context) {
+    return copyFilesToInternalStorage(uri, context, "");
+  }
+
+  private static String copyFileToInternalStorage(Uri uri, Context context, String cacheFolder) {
+    ArrayList<Uri> uris = new ArrayList<Uri>() {};
+    uris.add(uri);
+    ArrayList<String> file = copyFilesToInternalStorage(uris, context, cacheFolder);
+
+    return file.get(0);
+  }
+
+  @NonNull
+  public static ArrayList<String> copyFilesToInternalStorage(
+      ArrayList<Uri> uris, Context context, String cacheFolderName) {
+    ArrayList<String> absolutePaths = new ArrayList<>();
+
+    String newDirPath = context.getFilesDir() + "/" + cacheFolderName;
+    createDirectoryIfNotExists(newDirPath);
+
+    for (Uri uri : uris) {
+      String name =
+          getFileName(
+              uri, context, new String[] {OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE});
+      File output = getFile(context, newDirPath, uri, name);
+      absolutePaths.add(output.getAbsolutePath());
+    }
+
+    return absolutePaths;
+  }
+
+  private static File getFile(Context context, String newDirPath, Uri uri, String name) {
+    File output;
+    output = new File(newDirPath + "/" + name);
+    try {
+      InputStream inputStream = context.getContentResolver().openInputStream(uri);
+      FileOutputStream outputStream = new FileOutputStream(output);
+      int read;
+      int bufferSize = 1024;
+      final byte[] buffers = new byte[bufferSize];
+      while ((read = inputStream.read(buffers)) != -1) {
+        outputStream.write(buffers, 0, read);
+      }
+
+      inputStream.close();
+      outputStream.close();
+
+    } catch (Exception e) {
+      System.out.println("There was an error adding a file to the application cache");
+    }
+    return output;
+  }
+
+  private static void createDirectoryIfNotExists(String newDirPath) {
+    File dir = new File(newDirPath);
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+  }
+
+  /**
+   * Returns an absolute path for the file to be saved from the 3 possible content paths to return.
+   * If none of these match, then we will create a copy of the created file in the internal storage
```

Comment:
Related to the comment below: I'm not sure what "the created file" is referring to here. What is being created?

Reply:
I'll answer both questions here:



> Related to the comment below: I'm not sure what "the created file" is referring to here. What is being created?

This is related to [this issue](https://github.com/flutter/flutter/issues/111583) for iOS. Basically, for mobile, we need to set a save path for a user to save a file. However, we must create an empty file to be able to write in, and then return the file path.

If you think this is a foul functionality, we may not include it in this first version, as iOS did (or didn't), and then discuss what can be done about it.



> Why would selecting a save location involve copying files? If this is actually necessary, it definitely need an explanatory comment.
> 
> (Also, if for some reason we need to copy files, it shouldn't be done on the platform thread.)

First, copying files to the internal storage is:
 - Done in Android Q+
 - Not actual internal storage, but the cache of the app using this package

Android Q+ requires new files or access to files to be accessed from the cache. So this might be a polemic addition to the package.
However, we were thinking of this from a user perspective, where you want (using flutter in Android) to prompt a user to save, for example, an image, so it is an addition that might be useful in the future.

I agree to keep it out of the platform thread; if you think this functionality can be implemented now, we can think of another way. Otherwise, as I mentioned before, we can leave it as an issue to tackle in a post-deployment.

Reply:
> Basically, for mobile, we need to set a save path for a user to save a file. However, we must create an empty file to be able to write in, and then return the file path.
> 
> If you think this is a foul functionality, we may not include it in this first version, as iOS did (or didn't), and then discuss what can be done about it.

We should not include this implementation if it's that fundamentally different from how this works on other platforms. If/when there's actually demand for this functionality on mobile, we can evaluate an API that actually matches the platform semantics.

Reply:
Ok! I'll remove this and leave it as it is in iOS, and we could probably create an issue to track it in the future, what do you think?

Reply:
Making an issue (to capture the investigation done so far, and to provide people something to vote for if they want it) sounds good to me.

Reply:
[Created!](https://github.com/flutter/flutter/issues/113441)